File: 39.combination-sum.py
#
# @lc app=leetcode id=39 lang=python3
#
# [39] Combination Sum
#

# @lc code=start
class Solution:
    r = list()
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        l = list()
        l.append(candidates[0])
        self.recursion(candidates, 0, l, target)
        return self.r
    
    def recursion(self, c, st, l, t):
        for i in range(st, len(c)):
            l.append(c[i])
            sum = 0
            for j in l: 
                sum+=j  # 0 3 0 1 6 8
            if sum==t:
                print(l, sum, t)
                if l not in self.r:
                    self.r.append(l)
                l = list()
                sum = 0
                continue
            if sum<t:
                self.recursion(c,i,l,t)
            else:
                if len(l)>1:
                    l=l[:len(l)-2]
            
# @lc code=end

File: 39.combination-sum.java
/*
 * @lc app=leetcode id=39 lang=java
 *
 * [39] Combination Sum
 */

// @lc code=start
import java.util.*;
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        combinationSum(0, candidates, target, new ArrayList<>(), result);
        return result;
    }
    private void combinationSum(int i, int[] candidates, int target, List<Integer> comb, List<List<Integer>> result) {
        if (target == 0) {
            result.add(new ArrayList<>(comb));
        }
        else{
            for (int j = i; j < candidates.length; j++) {
                if (candidates[j] > target) continue;
                comb.add(candidates[j]);
                combinationSum(j, candidates, target - candidates[j], comb, result);
                comb.remove(comb.size() - 1);
            }
        }
    }
}
// @lc code=end

